authorization/v2/uniformResourceIdentifier.go

```diff
@@ -24,6 +24,64 @@ func NewUriFromRequest(resource resources.Resource, handler gin.HandlerFunc, ctx
 	}
 }
 
+// NewURIFromString parses the string representation of a URI into the UniformResourceIdentifier struct.
+// NewURIFromString expects the string to be of the following form, otherwise ErrInvalidURI is returned.
+// hs:<service_name>:<subsystem>:<version>:<category>:<resource_name>?<allowed_arguments>#<permission_metadata>
+func NewURIFromString(source string) (UniformResourceIdentifier, error) {
+	var (
+		arguments map[string]string
+		metadata  map[string]string
+	)
+
+	pathArgumentSplit := strings.Split(source, "?")
+	path := pathArgumentSplit[0]
+
+	if len(pathArgumentSplit) > 2 {
+		return UniformResourceIdentifier{}, ErrInvalidURI
+	}
+
+	if len(pathArgumentSplit) == 2 {
+		var err error
+
+		argumentsMetadataSplit := strings.Split(pathArgumentSplit[1], "#")
+		arguments, err = unmarshalArguments(argumentsMetadataSplit[0])
+		if err != nil {
+			return UniformResourceIdentifier{}, err
```

Comment:
Could you maybe update the error handling so that it works similar to the way it does here: https://github.com/unicsmcr/hs_auth/blob/master/authorization/v2/authorizer.go

The public errors should only be returned by the public functions. Private functions should return private errors that the public functions then use to wrap the public errors.

So for example:
line 124 would be: `return nil, errors.New("malformed key value pair at index 1")`
and then line 49 would be: `return UniformResourceIdentifier{}, errors.Wrap(ErrInvalidURI, errors.Wrap(err, "could not unmarshall uri arguments"))`

The benefit of doing this is that we could still treat the error returned by `NewURIFromString` as an `ErrInvalidURI` like here: https://github.com/unicsmcr/hs_auth/blob/master/routers/api/v2/router.go#L104-L108 but when we print it out to the logs it would look like: `invalid URI string: could not unmarshall uri arguments: malformed key value pair at index 1` rather than just `invalid URI string`

Reply:
I'll look into refactoring the errors to follow this pattern 👍